File: annotation-processor/src/test/java/me/bristermitten/mittenlib/annotations/parser/ConfigClassParserTest.java
package me.bristermitten.mittenlib.annotations.parser;

import static org.assertj.core.api.AssertionsForClassTypes.assertThat;
import static org.assertj.core.api.AssertionsForInterfaceTypes.assertThatCollection;

import com.google.inject.Guice;
import com.palantir.javapoet.ClassName;
import com.palantir.javapoet.JavaFile;
import io.toolisticon.aptk.common.ToolingProvider;
import io.toolisticon.cute.Cute;
import io.toolisticon.cute.PassIn;
import javax.lang.model.element.TypeElement;
import me.bristermitten.mittenlib.annotations.ast.AbstractConfigStructure;
import me.bristermitten.mittenlib.annotations.compile.ConfigImplGenerator;
import me.bristermitten.mittenlib.annotations.compile.ConfigProcessorModule;
import me.bristermitten.mittenlib.annotations.integration.AtomicConfig;
import me.bristermitten.mittenlib.annotations.integration.InterfaceConfig;
import me.bristermitten.mittenlib.annotations.integration.IntersectionConfig;
import me.bristermitten.mittenlib.annotations.integration.UnionConfig;
import me.bristermitten.mittenlib.config.Config;
import me.bristermitten.mittenlib.config.names.ConfigName;
import org.assertj.core.api.InstanceOfAssertFactories;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

class ConfigClassParserTest {

    @Test
    void testParsingAtomicConfig() {
        Cute.unitTest()
                .when()
                .passInElement()
                .<TypeElement>fromClass(AtomicConfig.class)
                .intoUnitTest((processingEnvironment, element) -> {
                    var injector = Guice.createInjector(new ConfigProcessorModule(processingEnvironment));

                    AbstractConfigStructure ast =
                            injector.getInstance(ConfigClassParser.class).parseAbstract(element);

                    assertThat(ast).isNotNull().isInstanceOf(AbstractConfigStructure.Atomic.class);
                })
                .thenExpectThat()
                .compilationSucceeds()
                .executeTest();
    }

    @Test
    void testParsingIntersectionConfig() {
        Cute.unitTest()
                .when()
                .passInElement()
                .<TypeElement>fromClass(IntersectionConfig.class)
                .intoUnitTest((processingEnvironment, element) -> {
                    var injector = Guice.createInjector(new ConfigProcessorModule(processingEnvironment));
                    AbstractConfigStructure ast =
                            injector.getInstance(ConfigClassParser.class).parseAbstract(element);

                    assertThat(ast).isNotNull();
                    assertThat(ast).isNotNull().isInstanceOf(AbstractConfigStructure.Atomic.class);

                    assertThatCollection(ast.enclosed())
                            .singleElement()
                            //                            .extracting(ConfigTypeAST::structure)
                            .isInstanceOf(AbstractConfigStructure.Intersection.class)
                            .extracting(AbstractConfigStructure::name)
                            .isEqualTo(ClassName.get(IntersectionConfig.ChildIntersectionConfig.class));
                })
                .thenExpectThat()
                .compilationSucceeds()
                .executeTest();
    }

    @Test
    void testParsingNestedConfig() {
        Cute.unitTest()
                .when()
                .passInElement()
                .<TypeElement>fromClass(InterfaceConfig.class)
                .intoUnitTest((processingEnvironment, element) -> {
                    var injector = Guice.createInjector(new ConfigProcessorModule(processingEnvironment));
                    AbstractConfigStructure ast =
                            injector.getInstance(ConfigClassParser.class).parseAbstract(element);

                    assertThat(ast).isNotNull().isInstanceOf(AbstractConfigStructure.Atomic.class);

                    assertThatCollection(ast.enclosed()).first().isInstanceOf(AbstractConfigStructure.Atomic.class);
                })
                .thenExpectThat()
                .compilationSucceeds()
                .executeTest();
    }

    @Test
    void testParsingUnionConfig() {
        Cute.unitTest()
                .when()
                .passInElement()
                .<TypeElement>fromClass(UnionConfig.class)
                .intoUnitTest((processingEnvironment, element) -> {
                    var injector = Guice.createInjector(new ConfigProcessorModule(processingEnvironment));
                    AbstractConfigStructure ast =
                            injector.getInstance(ConfigClassParser.class).parseAbstract(element);

                    assertThat(ast).isNotNull();
                    assertThat(ast)
                            .isNotNull()
                            .asInstanceOf(InstanceOfAssertFactories.type(AbstractConfigStructure.Union.class))
                            .extracting(AbstractConfigStructure.Union::alternatives)
                            .asInstanceOf(InstanceOfAssertFactories.list(AbstractConfigStructure.class))
                            .hasSize(2)
                            .first()
                            .isInstanceOf(
                                    AbstractConfigStructure.Intersection.class) // since it extends the parent type
                            .isNotNull();
                })
                .thenExpectThat()
                .compilationSucceeds()
                .executeTest();
    }

    @Test
    void testParsingBasicConfig() {
        Cute.unitTest()
                .when()
                .passInElement()
                .<TypeElement>fromClass(TestInterfaceConfig.class)
                .intoUnitTest((processingEnvironment, element) -> {
                    ToolingProvider.setTooling(processingEnvironment);
                    var injector = Guice.createInjector(new ConfigProcessorModule(processingEnvironment));

                    AbstractConfigStructure ast =
                            injector.getInstance(ConfigClassParser.class).parseAbstract(element);

                    //                    assertThat(ast)
                    //                            .extracting(ConfigTypeAST::structure)
                    //                            .extracting(ConfigTypeAST.ConfigStructure::name)
                    //                            .extracting(ClassName::simpleName)
                    //                            .isEqualTo(TestInterfaceConfig.class.getSimpleName());
                    //
                    //                    assertThatList(ast.structure().properties())
                    //                            .hasSize(2);
                    //
                    //                    assertThatList(ast.structure().properties())
                    //                            .first()
                    //                            .extracting(Property::name)
                    //                            .isEqualTo("name");
                    //
                    //                    assertThatList(ast.structure().properties())
                    //                            .first()
                    //                            .extracting(Property::settings)
                    //                            .extracting(ASTSettings.PropertyASTSettings::configName)
                    //                            .extracting(ConfigName::value)
                    //                            .isEqualTo("thing-name");

                    var generator = injector.getInstance(ConfigImplGenerator.class);
                    JavaFile emit = generator.emit(ast);

                    assertThat(emit).isNotNull();
                    assertThat(emit.typeSpec().name()).isEqualTo("TestInterfaceConfigImpl");
                })
                .thenExpectThat()
                .compilationSucceeds()
                .executeTest();
    }

    @SuppressWarnings("unused")
    @Config
    @PassIn
    private interface TestInterfaceConfig {
        @ConfigName("thing-name")
        String name();

        int age();
    }

    @Test
    void testParsingConfigWithMultipleParents() {
        Cute.unitTest()
                .when()
                .passInElement()
                .<TypeElement>fromSourceString("me.bristermitten.mittenlib.tests.MultiParentConfigDTO", """
                        package me.bristermitten.mittenlib.tests;
                        import io.toolisticon.cute.PassIn;
                        import me.bristermitten.mittenlib.config.Config;

                        interface SomeInterface {}
                        class SuperClass {}

                        @Config
                        @PassIn
                        public class MultiParentConfigDTO extends SuperClass implements SomeInterface {
                            public int level;
                        }
                        """)
                .intoUnitTest((processingEnvironment, element) -> {
                    var injector = Guice.createInjector(new ConfigProcessorModule(processingEnvironment));
                    var parser = injector.getInstance(ConfigClassParser.class);
                    Assertions.assertThrows(IllegalArgumentException.class, () -> parser.parseAbstract(element));
                })
                .thenExpectThat()
                .compilationSucceeds()
                .executeTest();
    }

    @Test
    void testParsingConfigEnum() {
        Cute.unitTest()
                .when()
                .passInElement()
                .<TypeElement>fromSourceString("me.bristermitten.mittenlib.tests.EnumConfigDTO", """
                        package me.bristermitten.mittenlib.tests;
                        import io.toolisticon.cute.PassIn;
                        import me.bristermitten.mittenlib.config.Config;

                        @Config
                        @PassIn
                        public enum EnumConfigDTO {
                            VAL
                        }
                        """)
                .intoUnitTest((processingEnvironment, element) -> {
                    var injector = Guice.createInjector(new ConfigProcessorModule(processingEnvironment));
                    var parser = injector.getInstance(ConfigClassParser.class);
                    Assertions.assertThrows(IllegalArgumentException.class, () -> parser.parseAbstract(element));
                })
                .thenExpectThat()
                .compilationSucceeds()
                .executeTest();
    }
}

File: annotation-processor/src/test/scala/me/bristermitten/mittenlib/annotations/parser/ConfigClassParserTest.scala
package me.bristermitten.mittenlib.annotations.parser

import com.google.inject.Guice
import com.palantir.javapoet.ClassName
import io.toolisticon.aptk.common.ToolingProvider
import io.toolisticon.cute.Cute
import javax.lang.model.element.TypeElement
import me.bristermitten.mittenlib.annotations.compile.ConfigImplGenerator
import me.bristermitten.mittenlib.annotations.compile.ConfigProcessorModule
import me.bristermitten.mittenlib.annotations.domain.ConfigStructure
import me.bristermitten.mittenlib.annotations.domain.Property
import me.bristermitten.mittenlib.annotations.integration.AtomicConfig
import me.bristermitten.mittenlib.annotations.integration.InterfaceConfig
import me.bristermitten.mittenlib.annotations.integration.IntersectionConfig
import me.bristermitten.mittenlib.annotations.integration.UnionConfig
import org.assertj.core.api.Assertions.assertThat
import org.scalatest.funsuite.AnyFunSuite
import org.scalatest.matchers.should.Matchers
import scala.jdk.CollectionConverters.*

class ConfigClassParserTest extends AnyFunSuite with Matchers {

  test("testParsingAtomicConfig") {
    Cute
      .unitTest()
      .when()
      .passInElement()
      .fromClass[TypeElement](classOf[AtomicConfig])
      .intoUnitTest((processingEnvironment, element) => {
        val injector =
          Guice.createInjector(new ConfigProcessorModule(processingEnvironment))
        val ast = injector
          .getInstance(classOf[ConfigParser])
          .getParsedStructure(element)

        assertThat(ast).isNotNull()
        ast shouldBe a[ConfigStructure.Atomic]
      })
      .thenExpectThat()
      .compilationSucceeds()
      .executeTest()
  }

  test("testParsingIntersectionConfig") {
    Cute
      .unitTest()
      .when()
      .passInElement()
      .fromClass[TypeElement](classOf[IntersectionConfig])
      .intoUnitTest((processingEnvironment, element) => {
        val injector =
          Guice.createInjector(new ConfigProcessorModule(processingEnvironment))
        val ast = injector
          .getInstance(classOf[ConfigParser])
          .getParsedStructure(element)

        assertThat(ast).isNotNull()
        ast shouldBe a[ConfigStructure.Atomic]

        val enclosed = ast.enclosed.asJava
        assertThat(enclosed).hasSize(1)
        val firstEnclosed = enclosed.get(0)
        firstEnclosed shouldBe a[ConfigStructure.Intersection]
        assertThat(firstEnclosed.name).isEqualTo(
          ClassName.get(classOf[IntersectionConfig.ChildIntersectionConfig])
        )
      })
      .thenExpectThat()
      .compilationSucceeds()
      .executeTest()
  }

  test("testParsingNestedConfig") {
    Cute
      .unitTest()
      .when()
      .passInElement()
      .fromClass[TypeElement](classOf[InterfaceConfig])
      .intoUnitTest((processingEnvironment, element) => {
        val injector =
          Guice.createInjector(new ConfigProcessorModule(processingEnvironment))
        val ast = injector
          .getInstance(classOf[ConfigParser])
          .getParsedStructure(element)

        assertThat(ast).isNotNull()
        ast shouldBe a[ConfigStructure.Atomic]

        val enclosed = ast.enclosed.asJava
        assertThat(enclosed).hasSize(1)
        enclosed.get(0) shouldBe a[ConfigStructure.Atomic]
      })
      .thenExpectThat()
      .compilationSucceeds()
      .executeTest()
  }

  test("testParsingUnionConfig") {
    Cute
      .unitTest()
      .when()
      .passInElement()
      .fromClass[TypeElement](classOf[UnionConfig])
      .intoUnitTest((processingEnvironment, element) => {
        val injector =
          Guice.createInjector(new ConfigProcessorModule(processingEnvironment))
        val ast = injector
          .getInstance(classOf[ConfigParser])
          .getParsedStructure(element)

        assertThat(ast).isNotNull()
        ast shouldBe a[ConfigStructure.Union]
        val union = ast.asInstanceOf[ConfigStructure.Union]

        val alternatives = union.alternatives.asJava
        assertThat(alternatives).hasSize(2)
        val firstAlternative = alternatives.get(0)
        firstAlternative shouldBe a[ConfigStructure.Intersection]
      })
      .thenExpectThat()
      .compilationSucceeds()
      .executeTest()
  }

  test("testParsingBasicConfig") {
    Cute
      .unitTest()
      .when()
      .passInElement()
      .fromSourceString[TypeElement](
        "me.bristermitten.mittenlib.tests.TestInterfaceConfig",
        """
          |package me.bristermitten.mittenlib.tests;
          |import io.toolisticon.cute.PassIn;
          |import me.bristermitten.mittenlib.config.Config;
          |import me.bristermitten.mittenlib.config.names.ConfigName;
          |
          |@Config
          |@PassIn
          |public interface TestInterfaceConfig {
          |    @ConfigName("thing-name")
          |    String name();
          |
          |    int age();
          |}
          |""".stripMargin
      )
      .intoUnitTest((processingEnvironment, element) => {
        ToolingProvider.setTooling(processingEnvironment)
        val injector =
          Guice.createInjector(new ConfigProcessorModule(processingEnvironment))
        val ast = injector
          .getInstance(classOf[ConfigParser])
          .getParsedStructure(element)

        assertThat(ast.name.simpleName).isEqualTo("TestInterfaceConfig")

        val properties = ast.properties.asJava
        assertThat(properties).hasSize(2)

        val firstProp = properties.get(0)
        assertThat(firstProp.name).isEqualTo("name")
        assertThat(firstProp.configName.get).isEqualTo("thing-name")

        val generator = injector.getInstance(classOf[ConfigImplGenerator])
        val emit = generator.emit(ast)

        assertThat(emit).isNotNull()
        assertThat(emit.typeSpec().name()).isEqualTo("TestInterfaceConfigImpl")
      })
      .thenExpectThat()
      .compilationSucceeds()
      .executeTest()
  }

  test("testParsingConfigWithMultipleParents") {
    Cute
      .unitTest()
      .when()
      .passInElement()
      .fromSourceString[TypeElement](
        "me.bristermitten.mittenlib.tests.MultiParentConfigDTO",
        """
          |package me.bristermitten.mittenlib.tests;
          |import io.toolisticon.cute.PassIn;
          |import me.bristermitten.mittenlib.config.Config;
          |
          |interface SomeInterface {}
          |class SuperClass {}
          |
          |@Config
          |@PassIn
          |public class MultiParentConfigDTO extends SuperClass implements SomeInterface {
          |    public int level;
          |}
          |""".stripMargin
      )
      .intoUnitTest((processingEnvironment, element) => {
        val injector =
          Guice.createInjector(new ConfigProcessorModule(processingEnvironment))
        val parser = injector.getInstance(classOf[ConfigParser])
        intercept[IllegalArgumentException] {
          parser.getParsedStructure(element)
        }
      })
      .thenExpectThat()
      .compilationSucceeds()
      .executeTest()
  }

  test("testParsingConfigEnum") {
    Cute
      .unitTest()
      .when()
      .passInElement()
      .fromSourceString[TypeElement](
        "me.bristermitten.mittenlib.tests.EnumConfigDTO",
        """
          |package me.bristermitten.mittenlib.tests;
          |import io.toolisticon.cute.PassIn;
          |import me.bristermitten.mittenlib.config.Config;
          |
          |@Config
          |@PassIn
          |public enum EnumConfigDTO {
          |    VAL
          |}
          |""".stripMargin
      )
      .intoUnitTest((processingEnvironment, element) => {
        val injector =
          Guice.createInjector(new ConfigProcessorModule(processingEnvironment))
        val parser = injector.getInstance(classOf[ConfigParser])
        intercept[IllegalArgumentException] {
          parser.getParsedStructure(element)
        }
      })
      .thenExpectThat()
      .compilationSucceeds()
      .executeTest()
  }
}
